### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by phlax**
- ~~optimize directory children vfolder assocations~~
- only set permissions once
- select_relates for directory parents in views

### CODE REVIEW

File: pootle/apps/pootle_translationproject/views.py [hunk 1]
```diff
@@ -160,10 +156,20 @@ class TPDirectoryMixin(TPMixin):
     export_url_path = "pootle-tp-export"
     translate_url_path = "pootle-tp-translate"
 
+    @property
+    def object_related(self):
+        tp_prefix = (
+            "parent__" * self.kwargs.get("dir_path", "").count("/"))
+        return [
+            "%stranslationproject" % tp_prefix,
+            "%stranslationproject" % tp_prefix,
```
Comment (unho) [inline pootle/apps/pootle_translationproject/views.py]: Why is this line included twice?

Comment (phlax) [inline pootle/apps/pootle_translationproject/views.py]: to be sure?

il remove

**Comment by unho** [comment]
Would be great if you can include some explanation on the "Only set permissions once on the request" commit message.

**Comment by phlax** [comment]
k - ill amend

File: pootle/core/browser.py [hunk 2]
```diff
@@ -180,10 +176,18 @@ def get_children(directory):
...
-                   for child_dir in directory.child_dirs.live().iterator()]
-
-    stores = [make_store_item(child_store)
-              for child_store in directory.child_stores.live().iterator()]
+    directories = []
+    vfti_pks = vftis_for_children(directory).values_list("pk", flat=True)
+    directories = [
+        make_directory_item(
+            child_dir,
+            sort=("priority" if child_dir.pk in vfti_pks else ""))
```
Comment (unho) [inline pootle/core/browser.py]: This won't work as expected. `child_dir` is a `Directory` while `vftis_for_children()` return `VirtualFolderTreeItem` objects, so a `child_dir.pk` will never be in `vfti_pks`.

File: pootle/core/browser.py [hunk 3]
```diff
@@ -180,10 +176,18 @@ def get_children(directory):
...
+        for child_dir
+        in directory.child_dirs.live().iterator()]
+
+    stores = [
+        make_store_item(child_store)
+        for child_store in
+        directory.child_stores.live().iterator()]
```
Comment (unho) [inline pootle/core/browser.py]: This is the kind of cleanup I prefer to be on a separate commit when I am reviewing some commit that has actual changes.

**Comment by unho** [comment]
lgtm. Haven't tested for performance.

**Comment by julen** [comment]
After testing the before/after of this PR when accessing overview pages under different directory levels, the query count definitely goes down. The query time on average remains similar; sometimes it's slightly worse though (in the magnitude of a few milliseconds).

I might be looking at the wrong place, so @phlax could you help us clarify under which circumstances we should see the optimization kicking in?

**Comment by phlax** [comment]
hi @julen - yep there are 2 optmizations with this code, both relate to TP or descendant views.

the first optimization prevents the permission code from being called 2x.

the second optimization grabs the parent(s) info in the initial query.

I have had consistent results of a speedup of around 1/3

**Comment by julen** [comment]
The second is the one I was checking. So if I understand correctly this only applies to accessing an overview page for the first time, right? The subsequent reads should be served from the LRU cache. Does it also apply to subdirectories?

On a second look I can see the improvement for the root of TPs, for subdirectories I get mixed results, usually the previous code being slightly faster (again, in the magnitude of a few ms). Does this align with your results?

**Comment by phlax** [comment]
the LRU cache is on the lifetime of the view class so it wont be cached the next time you load the page

> On a second look I can see the improvement for the root of TPs, for subdirectories I get mixed results, usually the previous code being slightly faster (again, in the magnitude of a few ms). Does this align with your results?

no - as i said - i saw a consistent 1/3 improvement - although i didnt test the commits in isolation
